.crawler/tests/peering.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::time::{Duration, Instant};
+
+use pea2pea::Pea2Pea;
+use snarkos_crawler::crawler::{Crawler, Opts};
+use snarkos_integration::TestNode;
+
+const NUM_NODES: usize = 10;
+
+#[tokio::test]
+async fn basics() {
+    // tracing_subscriber::fmt::init();
+
+    // Prepare some collections we'll be using.
+    let mut test_nodes = Vec::with_capacity(NUM_NODES);
+    let mut test_node_addrs = Vec::with_capacity(NUM_NODES);
+
+    // Start the test nodes.
+    for _ in 0..NUM_NODES {
+        let test_node = TestNode::default().await;
+        let test_node_addr = test_node.node().listening_addr().unwrap();
+
+        test_nodes.push(test_node);
+        test_node_addrs.push(test_node_addr);
+    }
+
+    // Connect the test nodes into a linear topology.
+    for (node, prev_node_addr) in test_nodes.iter().skip(1).zip(&test_node_addrs) {
+        node.node().connect(*prev_node_addr).await.unwrap();
+    }
+
+    // Double-check the topology.
+    for (i, node) in test_nodes.iter().enumerate() {
+        if i == 0 || i == NUM_NODES - 1 {
+            assert_eq!(node.node().num_connected(), 1);
+        } else {
+            assert_eq!(node.node().num_connected(), 2);
+        }
+    }
+
+    // Start the crawler.
+    let opts = Opts {
+        node: "127.0.0.1:0".parse().unwrap(),
+    };
+    let crawler = Crawler::new(opts).await;
+
+    // "Seed" the crawler with the address of the first node.
+    crawler.known_network.add_node(test_node_addrs[0]);
+
+    // Initialize the crawler.
+    crawler.run_periodic_tasks();
+
+    let now = Instant::now();
+    loop {
```

Comment:
Could we use `wait_until!` here?

Reply:
This is _exactly_ `wait_until` :smile:; hmm I could try moving it around in `snarkos-integration` or find some other, better place for it; I'll do that tomorrow :ok_hand:.